modules/dag-shared/src/main/scala/org/tessellation/dag/validators/TransactionValidator.scala

```diff
@@ -0,0 +1,208 @@
+    signedTransaction.hasValidSignature.ifM(
+      signedTransaction.value.validNel.pure[F],
+      Async[F].pure(InvalidSourceSignature.invalidNel)
+    )
+
+  private def validateSourceBalances(
+    transactions: Seq[Transaction]
+  ): F[ValidationResult[Seq[Transaction]]] = {
+    def validate(source: Address, txs: Seq[Transaction]): F[ValidationResult[Seq[Transaction]]] =
+      getBalance(source).map { balance =>
+        val totalAmount = txs.map(tx => tx.amount.coerce + tx.fee.coerce).sum
+        val remaining = balance.coerce - totalAmount
+
+        remaining >= BigInt(0)
+      }.ifM(
+        txs.validNel.pure[F],
+        Async[F].pure(InsufficientSourceBalance(source).invalidNel)
+      )
+
+    transactions
+      .groupBy(_.source)
+      .toSeq
+      .traverse { case (source, txs) => validate(source, txs) }
+      .map(_.combineAll)
+  }
+
+  private def validateAddressInequality(tx: Transaction): ValidationResult[Transaction] =
+    if (tx.source != tx.destination)
+      tx.validNel
+    else
+      SourceAndDestinationAddressAreEqual(tx).invalidNel
+}
+
+object TransactionValidator {
+
+  type ValidationResult[A] = ValidatedNel[TransactionValidationError, A]
+
+  sealed trait TransactionValidationError extends NoStackTrace {
+    val errorMessage: String
+    override def getMessage: String = errorMessage
+  }
+
+  sealed trait IncorrectParentReference extends TransactionValidationError {
+    val parentReference: TransactionReference
+
+    val errorMessage: String =
+      s"Transaction has an incorrect parent reference: $parentReference."
+  }
+
+  case class SourceAndDestinationAddressAreEqual(source: Address, destination: Address)
+      extends TransactionValidationError {
+
+    val errorMessage: String =
+      s"Transaction's source and destination address are equal. Source=$source destination=$destination"
+  }
+
+  case class ParentTxRefOrdinalLowerThenStoredLastTxRef(parentReference: TransactionReference)
+      extends IncorrectParentReference
+
+  case class SameOrdinalButDifferentHashForLastTxRef(parentReference: TransactionReference)
+      extends IncorrectParentReference
+
+  case class NonZeroOrdinalButEmptyHash(parentReference: TransactionReference) extends IncorrectParentReference
+
+  case class SourceAddressAndSignerIdsDontMatch(source: Address) extends TransactionValidationError {
+    val errorMessage: String = s"Transactions source address=$source doesn't match the signing ids"
+  }
+
+  case class InsufficientSourceBalance(source: Address) extends TransactionValidationError {
+    val errorMessage: String = s"Transactions source address=$source has insufficient balance"
+  }
+
+  case object InvalidSourceSignature extends TransactionValidationError {
+    val errorMessage: String = s"Transaction has invalid source signature"
+  }
+
+  object SourceAddressAndSignerIdsDontMatch {
+    def apply(tx: Transaction) = new SourceAddressAndSignerIdsDontMatch(tx.source)
+  }
```

Comment:
I think you should order it like:
```
class AA
object AA

class BB
object BB
```
instead of 
```
class AA
class BB
object AA
object BB
```

Reply:
I will try, not sure if that wasn't re-ordered like that byt the linter (thought that could have been some internal intellij linter, so I will check that again).